packages/vkui/src/components/AppRoot/AppRoot.tsx

```diff
@@ -82,161 +87,91 @@ export const AppRoot = ({
-        documentBody.removeChild(portal);
-        portalRootRef.current = null;
-      };
-    },
-    [portalRootProp],
-  );
-
-  useIsomorphicLayoutEffect(
-    function setupContainers() {
-      if (!appRootRef.current || !portalRootRef.current) {
-        return;
-      }
-
-      const parentElement = getParentElement(appRootRef.current);
-      const documentBody = getDocumentBody(appRootRef.current);
-      const documentElement = documentBody.ownerDocument.documentElement;
 
-      const [baseClassNames, stylesClassNames] = getClassNamesByMode({
-        mode,
-        layout,
-        tokensClassName,
-        sizeX,
-        sizeY,
-      });
-
-      /* eslint-disable no-restricted-properties */
-      switch (mode) {
-        case 'full': {
-          if (parentElement) {
-            parentElement.classList.add(...baseClassNames);
-          }
-
-          documentElement.classList.add(...stylesClassNames, 'vkui');
-          const unsetSafeAreaInsets = setSafeAreaInsets(safeAreaInsets, documentElement);
-
-          return function cleanup() {
-            if (parentElement) {
-              parentElement.classList.remove(...baseClassNames);
-            }
-
-            documentElement.classList.remove(...stylesClassNames, 'vkui');
-            unsetSafeAreaInsets();
-          };
-        }
-        case 'embedded': {
-          if (parentElement) {
-            parentElement.classList.add(...baseClassNames, ...stylesClassNames);
-            if (!disableParentTransformForPositionFixedElements) {
-              parentElement.style.setProperty('transform', 'translate3d(0, 0, 0)');
-            }
-            const unsetSafeAreaInsets = setSafeAreaInsets(safeAreaInsets, parentElement, portalRootRef.current); // prettier-ignore
-            return function cleanup() {
-              parentElement.classList.remove(...baseClassNames, ...stylesClassNames);
-              if (!disableParentTransformForPositionFixedElements) {
-                parentElement.style.removeProperty('transform');
-              }
-              unsetSafeAreaInsets();
-            };
-          }
-          /* istanbul ignore next: node.parentElement может быть null, но такой кейс в теории невозможен */
-          return;
-        }
-        /* istanbul ignore next: не покрывается за счёт теста на <AppRoot mode="partial" /> */
-        case 'partial': {
-          return;
-        }
-      }
-      /* eslint-enable no-restricted-properties */
-    },
-    [
-      mode,
-      layout,
-      disableParentTransformForPositionFixedElements,
-      tokensClassName,
-      sizeX,
-      sizeY,
-      safeAreaInsets,
-    ],
-  );
-
-  const ScrollController = React.useMemo(
-    () => (scroll === 'contain' ? ElementScrollController : GlobalScrollController),
-    [scroll],
-  );
+  const isKeyboardInputActiveRef = useKeyboardInputTracker();
+  const safeAreaInsets = React.useMemo(() => safeAreaInsetsProp, [safeAreaInsetsProp]);
```

Comment:
safeAreaInsetsProp это объект который может каждый раз меняться, то есть это бесполезный код

```suggestion  
  const {
    top: safeAreaInsetsPropTop,
    left: safeAreaInsetsPropRight,
    bottom: safeAreaInsetsPropBottom,
    left: safeAreaInsetsPropLeft,
  } = safeAreaInsetsProp ?? {};
  const safeAreaInsets = React.useMemo(
    () => ({
      top: safeAreaInsetsPropTop,
      right: safeAreaInsetsPropRight,
      bottom: safeAreaInsetsPropBottom,
      left: safeAreaInsetsPropLeft,
    }),
    [
      safeAreaInsetsPropBottom,
      safeAreaInsetsPropLeft,
      safeAreaInsetsPropRight,
      safeAreaInsetsPropTop,
    ],
  );
```

Или пока вернуть назад

```suggestion
  const safeAreaInsets = useObjectMemo(safeAreaInsetsProp);
```

Reply:
Если и правда постоянно меняется, то ты прав. Я, к стыду, иначе воспринимал.
Перенёс использование useMemo в helpers хук, чтомы место в AppRoot не занимать. ⚡ https://github.com/VKCOM/VKUI/pull/7739/commits/57d6a1d4732016163cdb7c95eb63363213aff8a0